### PR TITLE
fixed bug where findById with undefined returned the first doc

### LINKED ISSUE #3270: Should findById(undefined) return null?
Currently, findById(undefined) retrieves the first document. Intuitively I'd expect this to return null.

### PR DESCRIPTION
**Pull request by raydog**
This should fix #3270

### CODE REVIEW

**Comment by vkarpov15** [comment]
Worth discussing a little more, and this change is potentially backwards-breaking so we're gonna have to hold off on merging.

**Comment by raydog** [comment]
Delaying to think on this some more is understandable. My arguments in favor of changing this behavior:
- This behavior (undefined -> first doc) appeared between versions 4.0.1 and 4.0.2, so this patch actually brings the library back to the way it used to work a few months ago. undefined -> null was the behavior for years, for every version of the library that I checked in both the 3.x series, and the 2.x series.
- Nothing about this behavior sounds intentional. If you asked programmer who was new to this library what findById does when id is undefined, they would probably not answer 'return the first doc in the collection'. That's just bizarre, and you have to know about internal implementation details of the library to predict it.
- By this definition of 'backward breaking', any bugfix could be argued to be backward breaking. This behavior isn't documented anywhere, so changing it isn't breaking anything that was defined behavior to start with. Also, as noted earlier, this was the behavior of the library for years.
- This behavior is begging for subtle bugs. If someone is doing something like:

    ```javascript
    models.SomeModel.findById(this.optionalValue, cb);
    ```
  This would result in the wrong document being returned, which could be devastating should that document contain personal info.

Anyways, thanks for the consideration, and for the quick response. :)

**Comment by raydog** [comment]
Any updates on this? The more I think about it, the more I disagree with the 'backwards-breaking' label, as this behavior is a clear regression in the library, and a dangerous one at that. For perspective, I checked what the behavior was in different versions of Mongoose:

Mongoose Version| Item returned when `findById(undefined)`
-------|---------------
4.1.3  | Wrong document
4.1.2  | Wrong document
4.1.1  | Wrong document
4.1.0  | Wrong document
4.0.8  | Wrong document
4.0.7  | Wrong document
4.0.6  | Wrong document
4.0.5  | Wrong document
4.0.4  | Wrong document
4.0.3  | Wrong document
4.0.2  | Wrong document
4.0.1  | null
4.0.0  | null
4.0.0  | null
4.0.0  | null
4.0.0  | null
4.0.0  | null
4.0.0  | null
3.9.7  | null
3.9.6  | null
3.9.5  | null
3.9.4  | null
3.9.3  | null
3.9.2  | null
3.9.1  | null
3.9.0  | null
3.8.35 | null
3.8.34 | null
3.8.33 | null
3.8.31 | null
3.8.30 | null
3.8.29 | null
3.8.28 | null
3.8.27 | null
3.8.26 | null
3.8.25 | null
3.8.24 | null
3.8.23 | null
3.8.22 | null
3.8.21 | null
3.8.20 | null
3.8.19 | null
3.8.18 | null
3.8.17 | null
3.8.16 | null
3.8.15 | null
3.8.14 | null
3.8.13 | null
3.8.12 | null
3.8.11 | null
3.8.10 | null
3.8.9  | null
3.8.3  | null
3.8.2  | null
3.8.1  | null
3.8.0  | null
3.7.4  | null
3.7.3  | null
3.7.2  | null
3.7.2  | null
3.7.0  | null
3.6.20 | null
3.6.19 | null
3.6.18 | null
3.6.17 | null
3.6.16 | null
3.6.15 | null
3.6.14 | null
3.6.13 | null
3.6.12 | null
3.6.11 | null
3.6.10 | null
3.6.9  | null
3.6.8  | null
3.6.7  | null
3.6.6  | null
3.6.5  | null
3.6.4  | null
3.6.3  | null
3.6.2  | null
3.6.2  | null
3.6.2  | null
3.6.2  | null
3.6.2  | null
3.5.16 | null
3.5.15 | null
3.5.14 | null
3.5.13 | null
3.5.12 | null
3.5.11 | null
3.5.10 | null
3.5.9  | null
3.5.8  | null
3.5.7  | null
3.5.6  | null
3.5.5  | null
3.5.4  | null
3.5.3  | null
3.5.2  | null
3.5.1  | null
3.5.0  | null
3.4.0  | null
3.3.1  | null
3.3.0  | null
3.2.2  | null
3.2.1  | null
3.2.0  | null
3.2.0  | null
3.1.1  | null
3.1.0  | null
3.0.3  | null
3.0.2  | null
3.0.1  | null
3.0.0  | null
3.0.0  | null
3.0.0  | null
3.0.0  | null
2.7.3  | null
2.7.2  | null
2.7.1  | null
2.7.0  | null
2.6.8  | null
2.6.7  | null
2.6.6  | null
2.6.5  | null
2.6.4  | null
2.6.3  | null
2.6.2  | null
2.6.1  | null
2.6.0  | null
2.5.14 | null
2.5.13 | null
2.5.12 | null
2.5.11 | null
2.5.10 | null
2.5.9  | null
2.5.8  | null
2.5.7  | null
2.5.6  | null
2.5.5  | null
2.5.4  | null
2.5.3  | null
2.5.2  | null
2.5.1  | null
2.5.0  | null
2.4.10 | null
2.4.9  | null
2.4.8  | null
2.4.7  | null
2.4.6  | null
2.4.5  | null
2.4.4  | null
2.4.3  | null
2.4.2  | null
2.4.1  | null
2.4.0  | null
2.3.13 | null
2.3.12 | null
2.3.11 | null
2.3.10 | null
2.3.9  | null
2.3.8  | null
2.3.7  | null
2.3.6  | null
2.3.5  | null
2.3.4  | null
2.3.3  | null
2.3.2  | null
2.3.1  | null
2.3.0  | null
2.2.4  | null
2.2.3  | null
2.2.2  | null
2.2.1  | null
2.2.0  | null
2.1.4  | null
2.1.3  | null
2.1.2  | null
2.1.1  | null
2.1.0  | null
2.0.4  | null
2.0.3  | null
2.0.2  | null
2.0.1  | null
2.0.0  | null
1.8.4  | null
1.8.3  | null
1.8.2  | null
1.8.1  | null
1.8.0  | null
1.7.4  | null
1.7.3  | null
1.7.2  | null
1.7.2  | null
1.7.2  | null
1.6.0  | null
1.5.0  | null
1.4.0  | null
1.3.7  | null
1.3.6  | null
1.3.5  | null
1.3.4  | null
1.3.3  | null
1.3.2  | null
1.3.1  | null

File: lib/model.js [hunk 1]
```diff
@@ -1081,7 +1081,7 @@ Model.find = function find (conditions, projection, options, callback) {
  */
 
 Model.findById = function findById (id, projection, options, callback) {
-  return this.findOne({ _id: id }, projection, options, callback);
+  return this.findOne({ _id: id || null }, projection, options, callback);
```
Comment (vkarpov15) [inline lib/model.js]: There's a small problem with this; nobody will be able to query for documents with `_id` 0, false, etc.

**Pull request merged**